module/zfs/refcount.c

```diff
@@ -156,16 +159,14 @@ zfs_refcount_remove_many(zfs_refcount_t *rc, uint64_t number,
 	reference_t *ref;
 	int64_t count;
 
-	mutex_enter(&rc->rc_mtx);
-	ASSERT3U(rc->rc_count, >=, number);
-
 	if (!rc->rc_tracked) {
-		rc->rc_count -= number;
-		count = rc->rc_count;
-		mutex_exit(&rc->rc_mtx);
+		count = atomic_add_64_nv(&(rc)->rc_count, -number);
+		ASSERT3S(count, >=, 0);
 		return (count);
 	}
 
+	mutex_enter(&rc->rc_mtx);
+	ASSERT3U(rc->rc_count, >=, number);
```

Comment:
```suggestion
	ASSERT3U(zfs_refcount_count(rc), >=, number);
```

It's under the mutex but it still seems prudent to use `atomic_load_64` here.

Reply:
Why?  When tracking is enabled, all accesses are done under the mutex, so no atomics are needed.

Reply:
Indeed, my mistake.  LGTM.